pkg/engine/executor/parse.go

```diff
@@ -0,0 +1,117 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/apache/arrow-go/v18/arrow/array"
+	"github.com/apache/arrow-go/v18/arrow/memory"
+
+	"github.com/grafana/loki/v3/pkg/engine/internal/datatype"
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+	"github.com/grafana/loki/v3/pkg/engine/planner/logical"
+	"github.com/grafana/loki/v3/pkg/engine/planner/physical"
+)
+
+func NewParsePipeline(parse *physical.ParseNode, input Pipeline, allocator memory.Allocator) *GenericPipeline {
+	return newGenericPipeline(Local, func(ctx context.Context, inputs []Pipeline) state {
+		// Pull the next item from the input pipeline
+		input := inputs[0]
+		err := input.Read(ctx)
+		if err != nil {
+			return failureState(err)
+		}
+
+		batch, err := input.Value()
+		if err != nil {
+			return failureState(err)
+		}
+		// Batch needs to be released here since it won't be passed to the caller and won't be reused after
+		// this call to newGenericPipeline.
+		defer batch.Release()
+
+		// Find the message column
+		messageColIdx := -1
+		schema := batch.Schema()
+		for i := 0; i < schema.NumFields(); i++ {
+			if schema.Field(i).Name == types.ColumnNameBuiltinMessage {
+				messageColIdx = i
+				break
+			}
+		}
+		if messageColIdx == -1 {
+			return failureState(fmt.Errorf("message column not found"))
+		}
+
+		messageCol := batch.Column(messageColIdx)
+		stringCol, ok := messageCol.(*array.String)
+		if !ok {
+			return failureState(fmt.Errorf("message column is not a string column"))
+		}
+
+		// Parse logfmt based on the parser kind
+		var headers []string
+		var parsedColumns []arrow.Array
+		switch parse.Kind {
+		case logical.ParserLogfmt:
+			headers, parsedColumns = BuildLogfmtColumns(stringCol, parse.RequestedKeys, allocator)
+		default:
+			return failureState(fmt.Errorf("unsupported parser kind: %v", parse.Kind))
+		}
+
+		// Build new schema with original fields plus parsed fields
+		newFields := make([]arrow.Field, 0, schema.NumFields()+len(headers))
+		for i := 0; i < schema.NumFields(); i++ {
+			newFields = append(newFields, schema.Field(i))
+		}
+		for _, header := range headers {
+			// Add metadata to mark these as parsed columns
+			metadata := arrow.NewMetadata([]string{
+				types.MetadataKeyColumnType,
+				types.MetadataKeyColumnDataType,
+			}, []string{
+				types.ColumnTypeParsed.String(),
+				datatype.Loki.String.String(),
+			})
+			newFields = append(newFields, arrow.Field{
```

Comment:
`Nullable: true` should be set here since there's going to be at least some null values if the set of keys isn't consistent across rows.